wallet: #25768 follow ups

stickies-v opened the pull request:
This PR addresses the outstanding comments/issues from #25768:

- capitalization [typo](https://github.com/bitcoin/bitcoin/pull/25768#discussion_r958572522) in docstring
- remove [unused locks](https://github.com/bitcoin/bitcoin/commit/01f3534632d18c772901fb6ce22f6394eae96799) that we previously needed for `ReacceptWalletTransactions()`
- before #25768, only `ResendWalletTransactions()` would reset `m_next_resend` (formerly called `nNextResend`). By unifying it with `ReacceptWalletTransactions()` into `ResubmitWalletTransactions()`, the number of callsites that would reset the `m_next_resend` timer increased
  - since `m_next_resend` is only used in case of `relay=true` (formerly `ResendWalletTransactions()`), this is unintuitive
  - it leads to [unexpected behaviour](https://github.com/bitcoin/bitcoin/pull/25768#issuecomment-1252619427) such as transactions potentially never being rebroadcasted.
  - it makes the ResubmitWalletTransactions()` logic [more complicated than strictly necessary](https://github.com/bitcoin/bitcoin/pull/25768#discussion_r962828563)
  - since #25768, we relied on an earlier call of `ResubmitWalletTransactions(relay=false, force=true)` to initialize `m_next_resend()`, I think we can more elegantly do that by just providing `m_next_resend` with a default value
  - just to highlight: this commit introduces behaviour change

Note: the `if (!fBroadcastTransactions)` in `CWallet:ShouldResend()` is duplicated on purpose, since it potentially avoids the slightly more expensive `if (!chain().isReadyToBroadcast())` check afterwards. I don't have a strong view on it, so happy to remove that additional check to reduce the diff, too.

stickies-v committed:
refactor: remove unused locks for ResubmitWalletTransactions

ReacceptWalletTransactions is replaced by ResubmitWalletTransactions
which already handles acquiring the necessary locks internally.

stickies-v committed:
wallet: fix capitalization in docstring

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -537,6 +540,10 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     ScanResult ScanForWalletTransactions(const uint256& start_block, int start_height, std::optional<int> max_height, const WalletRescanReserver& reserver, bool fUpdate, const bool save_progress);
     void transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRemovalReason reason, uint64_t mempool_sequence) override;
+    /** Set the next time this wallet should resend transactions, or use the default. */
+    void SetNextResend(int next_resend = GetDefaultNextResend()) { m_next_resend = next_resend; }
```
Any reason to add a function that is only called once?

stickies-v replied:
Because `m_next_resend` is a private member, so I figured we'd prefer using a setter instead of making the variable public? Happy to take the other approach though, no strong preference.

MarcoFalke replied:
Sounds good. If you keep the function, the argument can be removed. If you keep the argument, the type needs to be adjusted to be not truncating.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -348,6 +349,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      */
     static bool AttachChain(const std::shared_ptr<CWallet>& wallet, interfaces::Chain& chain, const bool rescan_required, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
+    static int GetDefaultNextResend() { return GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60); }
```
Any reason to truncate the result?

stickies-v replied:
What do you mean with truncating here? I just moved the existing logic into a new function, that on its own shouldn't be any behavior change.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -250,7 +251,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     /** The next scheduled rebroadcast of wallet transactions. */
-    std::atomic<int64_t> m_next_resend{};
+    std::atomic<int64_t> m_next_resend{GetDefaultNextResend()};
```
No longer needs to be atomic?

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1979,7 +1984,9 @@ void CWallet::ResubmitWalletTransactions(bool relay, bool force)
     for (const std::shared_ptr<CWallet>& pwallet : GetWallets(context)) {
+        if (!pwallet->ShouldResend()) continue;
         pwallet->ResubmitWalletTransactions(/*relay=*/true, /*force=*/false);
+        pwallet->SetNextResend();
```
Might be good to have bugfixes in separate commits from refactoring changes

luke-jr replied:
The "bugfix" commit is full of refactoring still

luke-jr replied:
Also, cherry-picking the bugfix commit by itself is broken: it will reset `m_next_resend` to 12-36 hours, *every minute*, even if nothing is done/rebroadcast.

stickies-v committed:
refactor: revert m_next_resend to not be std::atomic

Since m_next_resend is now only called from MaybeResendWalletTxs()
we don't have any potential race conditions anymore, so the usage
of std::atomic can be reverted.

stickies-v committed:
refactor: carve out tx resend timer logic into ShouldResend

Moves the logic of whether or not transactions should actually be
resent out of the function that's resending them. This reduces
responsibilities of ResubmitWalletTransactions and allows
carving out the updating of m_next_resend in a future commit.

stickies-v committed:
wallet: only update m_next_resend when actually resending

We only want to relay our resubmitted transactions once every 12-36h.
By separating the timer update logic out of ResubmitWalletTransactions
and into MaybeResendWalletTxs we avoid non-relay calls (previously in
the separate ReacceptWalletTransactions function) from resetting that
timer.

stickies-v reviewed (commented):
Force pushed to address MarcoFalke's review comments - thank you!

- Split out "wallet: move resend timer logic out of ResubmitWalletTransactions" into two commits: one refactoring commit (that introduces `ShouldResend()` but has no behaviour change) and one bugfix commit that only updates `m_next_resend` when we're actually resending the transactions
- Fixed int truncation bug
- No need for `m_next_resend` to be `std::atomic` anymore, since it's now only called from a single thread (`MaybeResendWalletTxs()`)
- Made `SetNextResend()` argument-less since we're never actually passing the optional argument, can easily add back later.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -348,6 +349,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      */
     static bool AttachChain(const std::shared_ptr<CWallet>& wallet, interfaces::Chain& chain, const bool rescan_required, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
+    static int64_t GetDefaultNextResend() { return GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60); }
```
nit: Could put the impl into the cpp file to avoid a header include in the header?

maflcko reviewed (approved):
lgtm. I wouldn't call the last commit a "performance" improvement. Just a code clarity change.

aureleoules commented on src/wallet/wallet.cpp:
```diff
@@ -1934,17 +1953,6 @@ void CWallet::ResubmitWalletTransactions(bool relay, bool force)
-    // that these are our transactions.
-    if (!force && GetTime() < m_next_resend) return;
-    // resend 12-36 hours from now, ~1 day on average.
-    m_next_resend = GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60);
```
By removing this line, calls such as `importaddress`, `importpubkey`, `importmulti` and `importdescriptors` with rescan enabled won't update the timeout because now the timeout is only reset in `MaybeResendWalletTxs` with `SetNextResend`, is this normal?

See https://github.com/bitcoin/bitcoin/blob/6610cc99a6a479ca22d272a1d583ab2a8422b7ae/src/wallet/rpc/backup.cpp#L471-L475

stickies-v replied:
Yes, this is intentional. Prior to #25768, we had two functions [`ReacceptWalletTransactions()`](https://github.com/achow101/bitcoin/blob/e191fac4f3c37820f0618f72f0a8e8b524531ab8/src/wallet/wallet.cpp#L1860) (which did not update/read `nNextResend`) and [`ResendWalletTransactions()`](https://github.com/achow101/bitcoin/blob/e191fac4f3c37820f0618f72f0a8e8b524531ab8/src/wallet/wallet.cpp#L1937) (which did update/read `nNextResend`, only called from `MaybeUpdateWalletTxs()`), which got merged into a single parameterized `ResubmitWalletTransactions()` (and updates/reads `m_next_resend`).

Since the purpose of `m_next_resend` is to prevent overly frequent _relaying_ of transactions, we should only update the timer when relaying is triggered, so this PR just undoes that behaviour change introduced in #25768.

I tried to summarize this rationale in the commit message of 3e5d3f8705154e09ecf3639f6d261595992fc5b5 - lmk if that message is unclear?

aureleoules replied:
Great, that makes sense since these RPC calls don't relay transactions. Your commit message is also clear with this in mind :+1:.

aureleoules reviewed (commented):
01f3534632d18c772901fb6ce22f6394eae96799: I verified that `ResubmitWalletTransactions` already locks the mutex so no need to lock it in the parent functions.

Left a comment, the rest looks good.

stickies-v commented on src/wallet/wallet.cpp:
```diff
@@ -1903,6 +1903,23 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
     return result;
 }
 
+bool CWallet::ShouldResend()
```
Good catch, thx. Used to have side effects in `ShouldResend()` in an earlier implementation and forgot to remove.

stickies-v commented:
Force pushed to make `CWallet::ShouldResend()` `const` - ty aureleoules!

aureleoules commented:
ACK b01682a812f0841170657708ef0e896b904fcd77

luke-jr commented:
>I verified that ResubmitWalletTransactions already locks the mutex so no need to lock it in the parent functions.

That's not necessarily a given. The parents may need to lock things for atomicity (with other changes/calls), or lock ordering consistency. (I didn't look at this particular case, just noting the general logic does not hold)

luke-jr commented:
>it leads to https://github.com/bitcoin/bitcoin/pull/25768#issuecomment-1252619427 such as transactions potentially never being rebroadcasted.

I think this bugfix should be split out and backported. Ideally doing something more minimal which can be refactored here.

achow101 commented:
ACK b01682a812f0841170657708ef0e896b904fcd77

fanquake commented:
Added to #26133 for backport.
